modules/core/include/opencv2/core/quaternion.hpp

```diff
@@ -0,0 +1,1949 @@
+            // add warning:
+        }
+        T cosTheta = v0.dot(v1);
+        T DOT_THRESHOLD = 0.995;
+        if (cosTheta > DOT_THRESHOLD)
+        {
+            return nlerp(v0, v1, t, true);
+        }
+
+        if (directChange && cosTheta < 0)
+        {
+            v0 = -v0;
+            cosTheta = -cosTheta;
+        }
+        T sinTheta = std::sqrt(1 - cosTheta * cosTheta);
+        T angle = atan2(sinTheta, cosTheta);
+        return (std::sin((1 - t) * angle) / (sinTheta) * v0 + std::sin(t * angle) / (sinTheta) * v1).normalize();
+    }
+
+
+    template <typename T>
+    inline Quat<T> Quat<T>::nlerp(const Quat<T> &q0, const Quat<T> &q1, const T t, bool assumeUnit)
+    {
+        Quat<T> v0(q0), v1(q1);
+        if (v1.dot(v0) < 0)
+        {
+            v0 = -v0;
+        }
+        if (assumeUnit)
+        {
+            return ((1 - t) * v0 + t * v1).normalize();
+        }
+        // add warning
+        v0 = v0.normalize();
+        v1 = v1.normalize();
+        return ((1 - t) * v0 + t * v1).normalize();
+    }
+
+
+    template <typename T>
+    inline bool Quat<T>::isNormal(T eps) const
+    {
+
+        double normVar = norm();
+        if ((normVar > 1 - eps) && (normVar < 1 + eps))
+            return true;
+        return false;
+    }
+
+    template <typename T>
+    inline void Quat<T>::assertNormal(T eps) const
+    {
+        if (!isNormal(eps))
+            CV_Error(Error::StsBadArg, "Quaternion should be normalized");
+    }
+
+
+    template <typename T>
+    inline Quat<T> Quat<T>::squad(const Quat<T> &q0, const Quat<T> &q1,
+                                const Quat<T> &q2, const Quat<T> &q3,
+                                const T t, bool assumeUnit,
+                                bool directChange)
+    {
+        Quat<T> v0(q0), v1(q1), v2(q2), v3(q3);
+        if (!assumeUnit)
+        {
+            v0 = v0.normalize();
+            v1 = v1.normalize();
+            v2 = v2.normalize();
+            v3 = v3.normalize();
+        }
+
+        Quat<T> c0 = slerp(v0, v3, t, assumeUnit, directChange);
+        Quat<T> c1 = slerp(v1, v2, t, assumeUnit, directChange);
+        return slerp(c0, c1, 2 * t * (1 - t), assumeUnit, directChange);
+    }
+
+    template <typename T>
+    Quat<T> Quat<T>::interPoint(const Quat<T> &q0, const Quat<T> &q1,
+                                const Quat<T> &q2, bool assumeUnit)
+    {
+        Quat<T> v0(q0), v1(q1), v2(q2);
+        if (!assumeUnit)
+        {
+            v0 = v0.normalize();
+            v1 = v1.normalize();
+            v2 = v2.normalize();
+            // add warning in inter
+        }
+        return v1 * cv::exp(-(cv::log(v1.conjugate() * v0, assumeUnit) + (cv::log(v1.conjugate() * v2, assumeUnit))) / 4);
+    }
+
+    template <typename T>
+    Quat<T> Quat<T>::spline(const Quat<T> &q0, const Quat<T> &q1, const Quat<T> &q2, const Quat<T> &q3, const T t, bool assumeUnit)
+    {
+        Quatd v0, v1, v2, v3;
```

Comment:
You can squash this to copy constructor calls, not default constructors and copy assignments.

Reply:
OK, thanks